controllers/base.go

```diff
@@ -186,17 +270,50 @@ func (bmrc *baseMachineReconcileContext) DeleteMachineWithDependencies() error {
 		return fmt.Errorf("removing Workflow: %w", err)
 	}
 
-	if err := bmrc.releaseHardware(); err != nil {
+	if err := bmrc.releaseHardware(hardware); err != nil {
 		return fmt.Errorf("releasing Hardware: %w", err)
 	}
 
+	return nil
+}
+
+func (bmrc *baseMachineReconcileContext) removeFinalizer() error {
 	controllerutil.RemoveFinalizer(bmrc.tinkerbellMachine, infrastructurev1.MachineFinalizer)
 
 	bmrc.log.Info("Patching Machine object to remove finalizer")
 
 	return bmrc.patch()
 }
 
+// ensureBMCJobCompletionForDelete ensures the machine power off BMCJob is completed.
+// Removes the machint finalizer to let machine delete.
+func (bmrc *baseMachineReconcileContext) ensureBMCJobCompletionForDelete(hardware *tinkv1.Hardware) error {
```

Comment:
the 'ForDelete` part of this name gives me pause. With this in the name i should be able to infer that this method has some behavior around the deletion of a machine (tinkerbellMachine?). I dont see anything in the code that touches machine deletion.

Reply:
oh actually the `return bmrc.removeFinalizer()` is actually what allows the tinkerbellmachine objects to get deleted.
That's why i had it on the name.
We check job completion and then remove the finalizer.